akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala

```diff
@@ -314,33 +314,40 @@ object Behavior {
    */
-  def interpretMessage[T](behavior: Behavior[T], ctx: ActorContext[T], msg: T): Behavior[T] =
-    interpret(behavior, ctx, msg)
+  def interpretMessage[T](behavior: Behavior[T], ctx: ActorContext[T], msg: T): Behavior[T] = {
+    // optimization: duplicated logic with interpretSignal to avoid having to do the extra
+    // message typecheck in the happy case
+    behavior match {
+      case ext: ExtensibleBehavior[T] ⇒
+        val possiblyDeferredResult = ext.receive(ctx, msg.asInstanceOf[T])
+        start(possiblyDeferredResult, ctx)
+      case s: StoppedBehavior[T] ⇒ s
+      case EmptyBehavior         ⇒ UnhandledBehavior.asInstanceOf[Behavior[T]]
+      case IgnoreBehavior        ⇒ SameBehavior.asInstanceOf[Behavior[T]]
+      case _: UntypedPropsBehavior[_] ⇒
+        throw new IllegalArgumentException(s"cannot wrap behavior [$behavior] in " +
+          "Behaviors.setup, Behaviors.supervise or similar")
+      case d: DeferredBehavior[_] ⇒ throw new IllegalArgumentException(s"deferred [$d] should not be passed to interpreter")
+    }
+  }
 
   /**
    * Execute the behavior with the given signal
    */
-  def interpretSignal[T](behavior: Behavior[T], ctx: ActorContext[T], signal: Signal): Behavior[T] =
-    interpret(behavior, ctx, signal)
-
-  private def interpret[T](behavior: Behavior[T], ctx: ActorContext[T], msg: Any): Behavior[T] = {
+  def interpretSignal[T](behavior: Behavior[T], ctx: ActorContext[T], signal: Signal): Behavior[T] = {
+    // optimization: duplicated logic with interpretMessage to avoid having to do the extra
+    // message typecheck in the happy case
     behavior match {
-      case null ⇒ throw new InvalidMessageException("[null] is not an allowed message")
-      case SameBehavior | UnhandledBehavior ⇒
-        throw new IllegalArgumentException(s"cannot execute with [$behavior] as behavior")
+      case ext: ExtensibleBehavior[T] ⇒
+        val possiblyDeferredResult = ext.receiveSignal(ctx, signal)
+        start(possiblyDeferredResult, ctx)
+      case s: StoppedBehavior[T] ⇒ s
+      case EmptyBehavior         ⇒ UnhandledBehavior.asInstanceOf[Behavior[T]]
+      case IgnoreBehavior        ⇒ SameBehavior.asInstanceOf[Behavior[T]]
       case _: UntypedPropsBehavior[_] ⇒
         throw new IllegalArgumentException(s"cannot wrap behavior [$behavior] in " +
           "Behaviors.setup, Behaviors.supervise or similar")
       case d: DeferredBehavior[_] ⇒ throw new IllegalArgumentException(s"deferred [$d] should not be passed to interpreter")
-      case IgnoreBehavior         ⇒ SameBehavior.asInstanceOf[Behavior[T]]
-      case s: StoppedBehavior[T]  ⇒ s
-      case EmptyBehavior          ⇒ UnhandledBehavior.asInstanceOf[Behavior[T]]
-      case ext: ExtensibleBehavior[T] ⇒
-        val possiblyDeferredResult = msg match {
-          case signal: Signal ⇒ ext.receiveSignal(ctx, signal)
-          case m              ⇒ ext.receive(ctx, m.asInstanceOf[T])
```

Comment:
is it this message type check that the duplication is trying to optimize?
I would like that we verify that this actually has a measurable impact, or if the major benefits are from the reordering.

Reply:
but wasn't that changing both things?
I'm interested in if we can avoid the duplication, if most of the gain is anyway from the reordering.

Reply:
ok, somewhat surprising